src/main/java/org/reactivestreams/servlet/RequestPublisher.java

```diff
@@ -80,66 +90,72 @@ private void maybeRead() {
     // Integer.MAX_VALUE demand) the loop above can starve all other tasks from running (including, for example,
     // cancel signals). So we limit the number of sequential reads that we do, and if we reach that limit, we resubmit
     // to do any further reads, but giving the opportunity for other tasks to execute.
-    if (reads == MAX_SEQUENTIAL_READS) {
+    if (readsLeft == 0) {
       mutex.execute(this::maybeRead);
     }
   }
 
-  private void handleError(Throwable t) {
+  private void handleError(final Throwable t) {
     if (running) {
       running = false;
       subscriber.onError(t);
       subscriber = null;
     }
   }
 
-  @Override
-  public void subscribe(Subscriber<? super ByteBuffer> subscriber) {
-    if (subscriber == null) {
-      throw new NullPointerException("Subscriber passed to subscribe must not be null");
+  private void handleOnComplete() {
+    if (running) {
+      running = false;
+      subscriber.onComplete();
+      subscriber = null;
     }
+  }
+
+  @Override
+  public void subscribe(final Subscriber<? super ByteBuffer> subscriber) {
+    Objects.requireNonNull(subscriber, "Subscriber passed to subscribe must not be null");
     mutex.execute(() -> {
       if (this.subscriber == null) {
         this.subscriber = subscriber;
-        inputStream.setReadListener(new Listener());
+        inputStream.setReadListener(new Listener()); // Do we need to set the read-listener *before* the first request for data?
```

Comment:
I don't think it matters due to the mutual exclusion - even a synchronous invocation of the listener or subscription from the setReadListener or onSubscribe methods won't be executed until both are set.

Reply:
Indeed. I was just thinking whether it would save a bit of spinning at first or not.

Reply:
Note that the listeners are quite lazy in servlet IO - they only do something after you invoke `isReady` and it returns false on the stream you're reading/writing to, so if you register it, it's never going to immediately invoke any callbacks.